# Spam Detector

In [191]:
import pandas as pd
from sklearn.model_selection import train_test_split
#from sklearn.metrics import accuracy_score, sklearn, recall_score, f1_score
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, roc_auc_score,\
                            confusion_matrix, classification_report, balanced_accuracy_score


## Retrieve the Data

The data is located at [https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv](https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv)

Dataset Source: [UCI Machine Learning Library](https://archive.ics.uci.edu/dataset/94/spambase)

Import the data using Pandas. Display the resulting DataFrame to confirm the import was successful.

In [146]:
# Import the data
data = pd.read_csv("https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv")
data.head()

word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
0            0.00               0.64           0.64           0.0   
1            0.21               0.28           0.50           0.0   
2            0.06               0.00           0.71           0.0   
3            0.00               0.00           0.00           0.0   
4            0.00               0.00           0.00           0.0   

   word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
0           0.32            0.00              0.00                0.00   
1           0.14            0.28              0.21                0.07   
2           1.23            0.19              0.19                0.12   
3           0.63            0.00              0.31                0.63   
4           0.63            0.00              0.31                0.63   

   word_freq_order  word_freq_mail  ...  char_freq_;  char_freq_(  \
0             0.00            0.00  ...         0.00        0.000   
1             0.00            0.94  ...         0.00        0.132   
2             0.64            0.25  ...         0.01        0.143   
3             0.31            0.63  ...         0.00        0.137   
4             0.31            0.63  ...         0.00        0.135   

   char_freq_[  char_freq_!  char_freq_$  char_freq_#  \
0          0.0        0.778        0.000        0.000   
1          0.0        0.372        0.180        0.048   
2          0.0        0.276        0.184        0.010   
3          0.0        0.137        0.000        0.000   
4          0.0        0.135        0.000        0.000   

   capital_run_length_average  capital_run_length_longest  \
0                       3.756                          61   
1                       5.114                         101   
2                       9.821                         485   
3                       3.537                          40   
4                       3.537                          40   

   capital_run_length_total  spam  
0                       278     1  
1                      1028     1  
2                      2259     1  
3                       191     1  
4                       191     1  

[5 rows x 58 columns]

In [147]:
data.shape

(4601, 58)

## Predict Model Performance

You will be creating and comparing two models on this data: a Logistic Regression, and a Random Forests Classifier. Before you create, fit, and score the models, make a prediction as to which model you think will perform better. You do not need to be correct! 

Write down your prediction in the designated cells in your Jupyter Notebook, and provide justification for your educated guess.

*Random Forest Classifier might perform well. The reason is most likely the data is non-linear*

## Split the Data into Training and Testing Sets

In [148]:
# Create the labels set `y` and features DataFrame `X`
X = data.drop(columns='spam')
y = data['spam']

In [149]:
# Check the balance of the labels variable (`y`) by using the `value_counts` function.
y.value_counts()

spam
0    2788
1    1813
Name: count, dtype: int64

In [151]:
# Split the data into X_train, X_test, y_train, y_test
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=42)
X_train.head()

word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
1173            0.00                0.0           0.00           0.0   
1954            0.67                0.0           0.00           0.0   
256             0.00                0.0           1.42           0.0   
3341            0.00                0.0           0.24           0.0   
4469            0.00                0.0           0.00           0.0   

      word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
1173           0.00            0.00              1.23                0.00   
1954           0.00            0.00              0.00                0.00   
256            0.71            0.00              0.00                0.71   
3341           0.09            0.04              0.00                0.00   
4469           0.00            0.00              0.00                0.00   

      word_freq_order  word_freq_mail  ...  word_freq_conference  char_freq_;  \
1173             0.00            0.00  ...                   0.0        0.000   
1954             0.00            0.00  ...                   0.0        0.000   
256              0.00            0.71  ...                   0.0        0.000   
3341             0.04            0.00  ...                   0.0        0.014   
4469             0.00            0.00  ...                   0.0        0.000   

      char_freq_(  char_freq_[  char_freq_!  char_freq_$  char_freq_#  \
1173        0.000          0.0        1.939        0.000        0.000   
1954        0.055          0.0        0.000        0.000        0.000   
256         0.000          0.0        0.931        0.000        0.000   
3341        0.148          0.0        0.014        0.044        0.007   
4469        0.000          0.0        0.000        0.000        0.000   

      capital_run_length_average  capital_run_length_longest  \
1173                       1.379                           8   
1954                       1.644                          13   
256                       12.055                         117   
3341                       2.112                          26   
4469                       1.500                           2   

      capital_run_length_total  
1173                        40  
1954                        74  
256                        217  
3341                      1223  
4469                         3  

[5 rows x 57 columns]

## Scale the Features

Use the `StandardScaler` to scale the features data. Remember that only `X_train` and `X_test` DataFrames should be scaled.

In [152]:
from sklearn.preprocessing import StandardScaler

# Create the StandardScaler instance
scalar = StandardScaler()

In [153]:
# Fit the Standard Scaler with the training data
scalar.fit(X_train)

StandardScaler()

In [154]:
# Scale the training data
X_train_scaled = scalar.transform(X_train)
X_train_scaled

array([[-0.33933373, -0.16214839, -0.55144324, ..., -0.11649103,
        -0.21175305, -0.41909144],
       [ 1.8777157 , -0.16214839, -0.55144324, ..., -0.10814021,
        -0.18841316, -0.35960578],
       [-0.33933373, -0.16214839,  2.2325961 , ...,  0.21993678,
         0.29705664, -0.10941609],
       ...,
       [-0.33933373, -0.16214839, -0.55144324, ..., -0.08642808,
        -0.20241709, -0.40334524],
       [-0.33933373,  0.02143809, -0.55144324, ..., -0.10902256,
        -0.18841316, -0.18639753],
       [-0.00843083, -0.16214839,  0.25239911, ..., -0.09960031,
        -0.1137255 ,  0.10403246]])

In [155]:
X_test_scaled = scalar.transform(X_test)
X_test_scaled

array([[-0.33933373, -0.16214839, -0.55144324, ..., -0.12843428,
        -0.2444289 , -0.48382584],
       [ 2.01007686, -0.16214839,  0.84057643, ..., -0.12742588,
        -0.23976092, -0.43308807],
       [-0.33933373, -0.16214839,  1.23269465, ..., -0.11835027,
        -0.21642103, -0.30886801],
       ...,
       [-0.1077017 ,  0.01345607,  1.05624145, ..., -0.05189034,
         0.09633355,  1.21501465],
       [-0.33933373, -0.16214839, -0.55144324, ..., -0.10952676,
        -0.23042497, -0.47507794],
       [-0.33933373, -0.16214839, -0.23774866, ..., -0.07788818,
        -0.11839347,  0.83710575]])

## Create and Fit a Logistic Regression Model

Create a Logistic Regression model, fit it to the training data, make predictions with the testing data, and print the model's accuracy score. You may choose any starting settings you like. 

In [156]:
# Train a Logistic Regression model and print the model score
from sklearn.linear_model import LogisticRegression
lr = LogisticRegression(random_state=1)
lr.fit(X_train_scaled, y_train)

LogisticRegression(random_state=1)

In [157]:
# Make and save testing predictions with the saved logistic regression model using the test data
y_pred_lr = lr.predict(X_test_scaled)
# Review the predictions
y_pred_lr

array([0, 0, 0, ..., 1, 0, 0])

In [158]:
# Calculate the accuracy score by evaluating `y_test` vs. `testing_predictions`.
accuracy_score(y_test, y_pred_lr)

0.9226759339704604

## Create and Fit a Random Forest Classifier Model

Create a Random Forest Classifier model, fit it to the training data, make predictions with the testing data, and print the model's accuracy score. You may choose any starting settings you like. 

In [159]:
# Train a Random Forest Classifier model and print the model score
from sklearn.ensemble import RandomForestClassifier
rf = RandomForestClassifier(n_estimators=150, random_state=1).fit(X_train_scaled, y_train)
rf

RandomForestClassifier(n_estimators=150, random_state=1)

In [160]:
# Make and save testing predictions with the saved logistic regression model using the test data
y_pred_rf = rf.predict(X_test_scaled)

# Review the predictions
y_pred_rf

array([0, 0, 0, ..., 1, 0, 0])

In [161]:
# Calculate the accuracy score by evaluating `y_test` vs. `testing_predictions`.
accuracy_score(y_test, y_pred_rf)

0.9626411815812337

## Evaluate the Models

Which model performed better? How does that compare to your prediction? Write down your results and thoughts in the following markdown cell.

Logistic Regression Accuracy = 0.9226759339704604  
Random Forest Classifier Accuracy = 0.9626411815812337

*The Random Forest Classifier has a higher accuracy, suggesting it performs better in terms of overall correct classifications.  
Created Random Forest Classiifer model by varying n_estimators values from 50 to 300.
Random Forest classifier model seem to give highest accuracy score when n_estimators=150.*

## Additional evaluation metrics

Calculate following metrics:
- Precision
- Recall
- F1 Score
- ROC AOC

In [204]:
def evaluate_model(y_test, y_pred):
    precision = precision_score(y_test, y_pred, pos_label=1)  #pos_label=1 indicates msg has been flagged as spam
    recall = recall_score(y_test, y_pred, pos_label=1)
    f1 = f1_score(y_test, y_pred, pos_label=1)
    return precision, recall, f1


In [180]:
lr_metrics = evaluate_model(y_test, y_pred_lr)
print("Logistic Regression: Precision={}, Recall={}, F1={}".format(*lr_metrics))

rf_metrics = evaluate_model(y_test, y_pred_rf)
print("Random Forest Classifier: Precision={}, Recall={}, F1={}".format(*rf_metrics))

Logistic Regression: Precision=0.9406392694063926, Recall=0.8673684210526316, F1=0.9025191675794085
Random Forest Classifier: Precision=0.98, Recall=0.9284210526315789, F1=0.9535135135135135


In [205]:
# Create a classification report for Logistic Regression model
print(classification_report(y_test, y_pred_lr, labels = [1, 0]))

              precision    recall  f1-score   support

           1       0.94      0.87      0.90       475
           0       0.91      0.96      0.94       676

    accuracy                           0.92      1151
   macro avg       0.93      0.91      0.92      1151
weighted avg       0.92      0.92      0.92      1151



In [206]:
# Create a classification report for Random Forest classification model
print(classification_report(y_test, y_pred_rf, labels = [1, 0]))

              precision    recall  f1-score   support

           1       0.98      0.93      0.95       475
           0       0.95      0.99      0.97       676

    accuracy                           0.96      1151
   macro avg       0.97      0.96      0.96      1151
weighted avg       0.96      0.96      0.96      1151



In [195]:
# Calculate the balanced accuracy score
print(f'Linear Regression balanced accuracy = {balanced_accuracy_score(y_test, y_pred_lr)}')
print(f'Random Forest Classification balanced accuracy = {balanced_accuracy_score(y_test, y_pred_rf)}')

Linear Regression balanced accuracy = 0.9144534412955465
Random Forest Classification balanced accuracy = 0.9575537215820616


In [184]:
# Confusion Matrix for Linear Regression Model
confusion_matrix(y_test, y_pred_lr, labels = [1,0])

array([[412,  63],
       [ 26, 650]])

In [185]:
# Confusion Matrix for Random Forest Classifier Model
confusion_matrix(y_test, y_pred_rf, labels = [1,0])

array([[441,  34],
       [  9, 667]])

In [207]:
# Calculate ROC AUC score for Logistic Regression Model
# Predict values with probabilities
pred_probas = lr.predict_proba(X_test)

# We only need the predictions for the 1 class; use a list comprehension to 
# gather the second value from each list
pred_probas_firsts = [prob[1] for prob in pred_probas]

# Calculate the roc_auc_score for Linear Regression Model
print(f'Linear Regression ROC AUC score = {roc_auc_score(y_test, pred_probas_firsts)}')

#lr_roc_auc = roc_auc_score(y_test, lr.predict_proba(X_test)[:, 1])

Linear Regression ROC AUC score = 0.7177919651199003


/opt/anaconda3/envs/DEV1/lib/python3.12/site-packages/sklearn/base.py:486: UserWarning: X has feature names, but LogisticRegression was fitted without feature names
  warnings.warn(


In [208]:
# Calculate ROC AUC score for Random Forest Classifier Model
# Predict values with probabilities
pred_probas = rf.predict_proba(X_test)

# We only need the predictions for the 1 class; use a list comprehension to 
# gather the second value from each list
pred_probas_firsts = [prob[1] for prob in pred_probas]

# Calculate the roc_auc_score for Random Fortest Model
print(f'Random Forest Classifier ROC AUC score = {roc_auc_score(y_test, pred_probas_firsts)}')

#rf_roc_auc = roc_auc_score(y_test, rf.predict_proba(X_test)[:, 1])

Random Forest Classifier ROC AUC score = 0.9280660230457801


/opt/anaconda3/envs/DEV1/lib/python3.12/site-packages/sklearn/base.py:486: UserWarning: X has feature names, but RandomForestClassifier was fitted without feature names
  warnings.warn(


#### After evaluating additional metrics it appears Random Forest model is a better fit.